cpp/src/aztec/plonk/composer/ultra_composer.cpp

```diff
@@ -2498,10 +2800,159 @@ void UltraComposer::process_ROM_array(const size_t rom_id, const size_t gate_off
+    for (size_t i = 0; i < ram_array.records.size(); ++i) {
+        const RamRecord& record = ram_array.records[i];
+
+        const auto index = record.index;
+        const auto value = get_variable(record.value_witness);
+        const auto index_witness = add_variable(fr((uint64_t)index));
+        const auto timestamp_witess = add_variable(record.timestamp);
+        const auto value_witness = add_variable(value);
+        RamRecord sorted_record{
+            .index_witness = index_witness,
+            .timestamp_witness = timestamp_witess,
+            .value_witness = value_witness,
+            .index = index,
+            .timestamp = record.timestamp,
+            .access_type = record.access_type,
+            .record_witness = 0,
+            .gate_index = 0,
+        };
+
+        // We don't apply the RAM consistency check gate to the final record,
+        // as this gate expects a RAM record to be present at the next gate
+        if (i < ram_array.records.size() - 1) {
+            create_sorted_RAM_gate(sorted_record);
+        } else {
+            // For the final record in the sorted list, we do not apply the full consistency check gate.
+            // Only need to check the index value = RAM array size - 1.
+            create_final_sorted_RAM_gate(sorted_record, ram_array.state.size());
+        }
+
+        // Assign record/sorted records to tags that we will perform set equivalence checks on
+        assign_tag(record.record_witness, access_tag);
+        assign_tag(sorted_record.record_witness, sorted_list_tag);
+
+        // For ROM/RAM gates, the 'record' wire value (wire column 4) is a linear combination of the first 3 wire
+        // values. However...the record value uses the random challenge 'eta', generated after the first 3 wires are
+        // committed to. i.e. we can't compute the record witness here because we don't know what `eta` is! Take the
+        // gate indices of the two rom gates (original read gate + sorted gate) and store in `memory_records`. Once we
+        // generate the `eta` challenge, we'll use `memory_records` to figure out which gates need a record wire value
+        // to be computed.
+
+        switch (record.access_type) {
+        case RamRecord::AccessType::READ: {
+            memory_read_records.push_back(
+                static_cast<uint32_t>(sorted_record.gate_index + gate_offset_from_public_inputs));
+            memory_read_records.push_back(static_cast<uint32_t>(record.gate_index + gate_offset_from_public_inputs));
+            break;
+        }
+        case RamRecord::AccessType::WRITE: {
+            memory_write_records.push_back(
+                static_cast<uint32_t>(sorted_record.gate_index + gate_offset_from_public_inputs));
+            memory_write_records.push_back(static_cast<uint32_t>(record.gate_index + gate_offset_from_public_inputs));
+            break;
+        }
+        default: {
+            ASSERT(false); // shouldn't get here!
```

Comment:
https://en.cppreference.com/w/cpp/utility/unreachable is coming, but for now maybe `__builtin_unreachable();`